### PR TITLE
Transformer models generation supports user-provided input embeddings

### PR DESCRIPTION
**Pull request by zongwave**
# What does this PR do?
Some multimodal generation model needs user specified embedded tokens input, for example [NExT-GPT](https://next-gpt.github.io/), this commit enable user-provided input embeddings for model generation.

1. Input Embeddings Support:

Modified transformers/generation/utils.py to include logic that supports user-provided input embeddings. This allows for more flexibility in how input data can be fed into the models.

Added the --input_embeds option in example/text-generation/run_generation.py to facilitate testing with different models using embedded tokens.

2. Testing and Performance Evaluation:

Conducted tests using the modified script on the Mistral 7B model with a batch size of 6. The tests included various input scenarios to assess the robustness and performance of the new features.

Enhanced example/text-generation/run_generation.py to support multiple decoding strategies including Greedy, Beam Search, and Contrastive Search. This allows users to specify the decoding strategy during runtime and evaluate the effectiveness of each method.

Fixes # (issue)
Optimum habana transformer models currently do not support user provided input embeddings for token generation.

Usage
To test the new features, use the following command:
```
python run_generation.py --model_name_or_path mistralai/Mistral-7B-Instruct-v0.2  --input_embeds --batch_size 6
```

Performance Data:

The table below summarizes the performance data for six popular models with both embeds and tokens input, highlighting the throughput, number of HPU graphs, graph compilation duration, and memory allocation.

<!DOCTYPE html>
<html>
<head>
    <style>
        body {
            font-family: "Intel Clear", sans-serif;
        }
        table {
            width: 100%;
            border-collapse: collapse;
        }
        th, td {
            border: 1px solid black;
            padding: 8px;
            text-align: left;
        }
        th {
            background-color: #f2f2f2;
        }
    </style>
</head>
<body>
    <table>
        <tr>
            <th>Model Name or Path</th>
            <th>Input Type</th>
            <th>Throughput (tokens/second)</th>
            <th>Number of HPU Graphs</th>
            <th>Graph Compilation Duration (seconds)</th>
            <th>Max Memory Allocated (GB)</th>
        </tr>
        <tr>
            <td rowspan="2">tiiuae/falcon-7b</td>
            <td>Embeds</td>
            <td>32.37</td>
            <td>8</td>
            <td>13.58</td>
            <td>26.97</td>
        </tr>
        <tr>
            <td>Tokens</td>
            <td>29.77</td>
            <td>7</td>
            <td>13.74</td>
            <td>26.96</td>
        </tr>
        <tr>
            <td rowspan="2">mosaicml/mpt-7b</td>
            <td>Embeds</td>
            <td>50.4</td>
            <td>8</td>
            <td>8.77</td>
            <td>25.68</td>
        </tr>
        <tr>
            <td>Tokens</td>
            <td>50.73</td>
            <td>7</td>
            <td>9.39</td>
            <td>25.68</td>
        </tr>
        <tr>
            <td rowspan="2">gpt2</td>
            <td>Embeds</td>
            <td>93.18</td>
            <td>8</td>
            <td>4.68</td>
            <td>0.65</td>
        </tr>
        <tr>
            <td>Tokens</td>
            <td>76.91</td>
            <td>7</td>
            <td>5.1</td>
            <td>0.65</td>
        </tr>
        <tr>
            <td rowspan="2">microsoft/phi-2</td>
            <td>Embeds</td>
            <td>31.19</td>
            <td>8</td>
            <td>14.62</td>
            <td>10.48</td>
        </tr>
        <tr>
            <td>Tokens</td>
            <td>29.55</td>
            <td>7</td>
            <td>14.63</td>
            <td>10.48</td>
        </tr>
        <tr>
            <td rowspan="2">mistralai/Mistral-7B-Instruct-v0.2</td>
            <td>Embeds</td>
            <td>29.44</td>
            <td>13</td>
            <td>11.04</td>
            <td>28.07</td>
        </tr>
        <tr>
            <td>Tokens</td>
            <td>25.6</td>
            <td>11</td>
            <td>12.09</td>
            <td>28.07</td>
        </tr>
        <tr>
            <td rowspan="2">Customized LLaMA-based Vicuna</td>
            <td>Embeds</td>
            <td>28.28</td>
            <td>12</td>
            <td>10.6</td>
            <td>25.31</td>
        </tr>
        <tr>
            <td>Tokens</td>
            <td>28.61</td>
            <td>11</td>
            <td>11.1</td>
            <td>25.31</td>
        </tr>
    </table>
</body>
</html>

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [Y] Did you write any new necessary tests?

### CODE REVIEW

**Comment by libinta** [comment]
@zongwave  as this PR changes the common utils.py, can you run through gaudi2 ci test for text-generation? 
https://github.com/huggingface/optimum-habana/blob/main/tests/test_text_generation_example.py
The way to trigger ci test
GAUDI2_CI=true RUN_SLOW=1 python -m pytest test/test_text-generation_example.py -m -k -v xxx
where xxx is specific model or test case,
please validate llama bf16/fp8, 1x/8x

**Comment by zongwave** [comment]
> @zongwave as this PR changes the common utils.py, can you run through gaudi2 ci test for text-generation? https://github.com/huggingface/optimum-habana/blob/main/tests/test_text_generation_example.py The way to trigger ci test GAUDI2_CI=true RUN_SLOW=1 python -m pytest test/test_text-generation_example.py -m -k -v xxx where xxx is specific model or test case, please validate llama bf16/fp8, 1x/8x

@libinta 
I ran GAUDI_CI test with command: GAUDI2_CI=1 RUN_SLOW=1 python -m pytest tests/test_text_generation_example.py -v -s -k meta-llama/Llama-2-7b-hf

The test passed with message:

PASSED

========================================================= 10 passed, 42 deselected in 2571.60s (0:42:51) ==========================================================

**Comment by zongwave** [comment]
> > @zongwave as this PR changes the common utils.py, can you run through gaudi2 ci test for text-generation? https://github.com/huggingface/optimum-habana/blob/main/tests/test_text_generation_example.py The way to trigger ci test GAUDI2_CI=true RUN_SLOW=1 python -m pytest test/test_text-generation_example.py -m -k -v xxx where xxx is specific model or test case, please validate llama bf16/fp8, 1x/8x
> 
> @libinta I ran GAUDI_CI test with command:
> 
> GAUDI2_CI=1 RUN_SLOW=1 python -m pytest tests/test_text_generation_example.py -v -k meta-llama/Llama-2-7b-hf ======================================================================= test session starts ======================================================================= platform linux -- Python 3.10.12, pytest-8.3.2, pluggy-1.5.0 -- /usr/bin/python cachedir: .pytest_cache rootdir: /workspace/src/optimum-habana configfile: setup.cfg plugins: anyio-4.4.0 collected 52 items / 42 deselected / 10 selected
> 
> tests/test_text_generation_example.py::test_text_generation_bf16_1x[token0-meta-llama/Llama-2-7b-hf-1-True-141.25776956002076] PASSED [ 10%] tests/test_text_generation_example.py::test_text_generation_bf16_1x[token0-meta-llama/Llama-2-7b-hf-512-True-12808] PASSED [ 20%] tests/test_text_generation_example.py::test_text_generation_bf16_1x[token0-meta-llama/Llama-2-7b-hf-512-False-8711] PASSED [ 30%] tests/test_text_generation_example.py::test_text_generation_fp8[token0-meta-llama/Llama-2-7b-hf-1-1230-False-128-128-13152.7] PASSED [ 40%] tests/test_text_generation_example.py::test_text_generation_fp8[token0-meta-llama/Llama-2-7b-hf-1-163-False-128-2048-4774.7] PASSED [ 50%] tests/test_text_generation_example.py::test_text_generation_fp8[token0-meta-llama/Llama-2-7b-hf-1-94-False-2048-128-1293.3] PASSED [ 60%] tests/test_text_generation_example.py::test_text_generation_fp8[token0-meta-llama/Llama-2-7b-hf-1-81-False-2048-2048-1942.9] PASSED [ 70%] tests/test_text_generation_example.py::test_text_generation_torch_compile[token0-meta-llama/Llama-2-7b-hf-102.27823420713148] PASSED [ 80%] tests/test_text_generation_example.py::test_text_generation_torch_compile_distributed[token0-meta-llama/Llama-2-7b-hf-39.72973199515235] PASSED [ 90%] tests/test_text_generation_example.py::test_text_generation_distributed_tp[token0-meta-llama/Llama-2-7b-hf-1345.2369318328463] PASSED [100%]
> 
> ========================================================= 10 passed, 42 deselected in 2522.21s (0:42:02) ==========================================================

I'm trying to figure out how to trigger llama 8x test case.

**Comment by regisss** [comment]
> I'm trying to figure out how to trigger llama 8x test case.

Llama 8x tests are [this one](https://github.com/huggingface/optimum-habana/blob/65d044e77692f01a299938ec06df0e5a6c4cb232/tests/test_text_generation_example.py#L55) and [this one](https://github.com/huggingface/optimum-habana/blob/65d044e77692f01a299938ec06df0e5a6c4cb232/tests/test_text_generation_example.py#L69).

**Comment by zongwave** [comment]
> > I'm trying to figure out how to trigger llama 8x test case.
> 
> Llama 8x tests are [this one](https://github.com/huggingface/optimum-habana/blob/65d044e77692f01a299938ec06df0e5a6c4cb232/tests/test_text_generation_example.py#L55) and [this one](https://github.com/huggingface/optimum-habana/blob/65d044e77692f01a299938ec06df0e5a6c4cb232/tests/test_text_generation_example.py#L69).

@regisss @libinta 
I ran Gaudi2_CI test with command:
# GAUDI2_CI=1 RUN_SLOW=1 python -m pytest tests/test_text_generation_example.py -v -k **llama**
18 llama cases were selected to be tested.

I triggered the CI test with "--input_embeds" option manually in test_text_generation_example.py
    command += [
        f"{path_to_example_dir / 'text-generation' / 'run_generation.py'}",
        f"--model_name_or_path {model_name}",
        f"--batch_size {batch_size}",
        "--use_kv_cache",
        **"--input_embeds"**,
        f"--max_new_tokens {max_output_tokens}",
    ]

**1. This is results of test on input embeds generation enabled by adding option "--input_embeds" manually:**
============================================================================================ test session starts =============================================================================================
platform linux -- Python 3.10.12, pytest-8.3.2, pluggy-1.5.0 -- /usr/bin/python
cachedir: .pytest_cache
rootdir: /workspace/src/optimum-habana
configfile: setup.cfg
plugins: anyio-4.4.0
collected 52 items / 34 deselected / 18 selected

tests/test_text_generation_example.py::test_text_generation_bf16_1x[token0-meta-llama/Llama-2-7b-hf-1-True-141.25776956002076] PASSED                                                                  [  5%]
tests/test_text_generation_example.py::test_text_generation_bf16_1x[token0-meta-llama/Meta-Llama-3-8B-1-True-129] PASSED                                                                               [ 11%]
tests/test_text_generation_example.py::test_text_generation_bf16_1x[token0-meta-llama/Llama-2-7b-hf-512-True-12808] PASSED                                                                             [ 16%]
tests/test_text_generation_example.py::test_text_generation_bf16_1x[token0-meta-llama/Llama-2-7b-hf-512-False-8711] PASSED                                                                             [ 22%]
tests/test_text_generation_example.py::test_text_generation_bf16_1x[token0-codellama/CodeLlama-34b-hf-1-True-32.644] PASSED                                                                            [ 27%]
tests/test_text_generation_example.py::test_text_generation_fp8[token0-meta-llama/Llama-2-7b-hf-1-1230-False-128-128-13152.7] PASSED                                                                   [ 33%]
tests/test_text_generation_example.py::test_text_generation_fp8[token0-meta-llama/Llama-2-7b-hf-1-163-False-128-2048-4774.7] PASSED                                                                    [ 38%]
tests/test_text_generation_example.py::test_text_generation_fp8[token0-meta-llama/Llama-2-7b-hf-1-94-False-2048-128-1293.3] PASSED                                                                     [ 44%]
tests/test_text_generation_example.py::test_text_generation_fp8[token0-meta-llama/Llama-2-7b-hf-1-81-False-2048-2048-1942.9] PASSED                                                                    [ 50%]
tests/test_text_generation_example.py::test_text_generation_fp8[token0-meta-llama/Llama-2-70b-hf-4-3042-False-128-128-5374.6] PASSED                                                                   [ 55%]
tests/test_text_generation_example.py::test_text_generation_fp8[token0-meta-llama/Llama-2-70b-hf-4-750-False-128-2048-7422.4] PASSED                                                                   [ 61%]
tests/test_text_generation_example.py::test_text_generation_fp8[token0-meta-llama/Llama-2-70b-hf-4-207-False-2048-128-568.5] PASSED                                                                    [ 66%]
tests/test_text_generation_example.py::test_text_generation_fp8[token0-meta-llama/Llama-2-70b-hf-8-172-False-2048-2048-4656.2] PASSED                                                                  [ 72%]
tests/test_text_generation_example.py::test_text_generation_deepspeed[token0-meta-llama/Llama-2-70b-hf-8-1-64.10514998902435] PASSED                                                                   [ 77%]
tests/test_text_generation_example.py::test_text_generation_deepspeed[token0-meta-llama/Meta-Llama-3-70B-Instruct-8-1-64] PASSED                                                                        [ 83%]
tests/test_text_generation_example.py::test_text_generation_torch_compile[token0-meta-llama/Llama-2-7b-hf-102.27823420713148] PASSED                                                                   [ 88%]
tests/test_text_generation_example.py::test_text_generation_torch_compile_distributed[token0-meta-llama/Llama-2-7b-hf-39.72973199515235] PASSED                                                        [ 94%]
tests/test_text_generation_example.py::test_text_generation_distributed_tp[token0-meta-llama/Llama-2-7b-hf-1345.2369318328463] PASSED                                                                  [100%]
========================================================================== 18 passed, 34 deselected in 7952.60s (2:12:32) ==========================================================================a

**2. this is the original tests for input token generation:**
# GAUDI2_CI=1 RUN_SLOW=1 python -m pytest tests/test_text_generation_example.py -v -k llama
====================================================================== test session starts =======================================================================
platform linux -- Python 3.10.12, pytest-8.3.2, pluggy-1.5.0 -- /usr/bin/python
cachedir: .pytest_cache
rootdir: /workspace/src/optimum-habana
configfile: setup.cfg
plugins: anyio-4.4.0
collected 52 items / 34 deselected / 18 selected

tests/test_text_generation_example.py::test_text_generation_bf16_1x[token0-meta-llama/Llama-2-7b-hf-1-True-141.25776956002076] PASSED                      [  5%]
tests/test_text_generation_example.py::test_text_generation_bf16_1x[token0-meta-llama/Meta-Llama-3-8B-1-True-129] PASSED                                   [ 11%]
tests/test_text_generation_example.py::test_text_generation_bf16_1x[token0-meta-llama/Llama-2-7b-hf-512-True-12808] PASSED                                 [ 16%]
tests/test_text_generation_example.py::test_text_generation_bf16_1x[token0-meta-llama/Llama-2-7b-hf-512-False-8711] PASSED                                 [ 22%]
tests/test_text_generation_example.py::test_text_generation_bf16_1x[token0-codellama/CodeLlama-34b-hf-1-True-32.644] PASSED                                [ 27%]
tests/test_text_generation_example.py::test_text_generation_fp8[token0-meta-llama/Llama-2-7b-hf-1-1230-False-128-128-13152.7] PASSED                       [ 33%]
tests/test_text_generation_example.py::test_text_generation_fp8[token0-meta-llama/Llama-2-7b-hf-1-163-False-128-2048-4774.7] PASSED                        [ 38%]
tests/test_text_generation_example.py::test_text_generation_fp8[token0-meta-llama/Llama-2-7b-hf-1-94-False-2048-128-1293.3] PASSED                         [ 44%]
tests/test_text_generation_example.py::test_text_generation_fp8[token0-meta-llama/Llama-2-7b-hf-1-81-False-2048-2048-1942.9] PASSED                        [ 50%]
tests/test_text_generation_example.py::test_text_generation_fp8[token0-meta-llama/Llama-2-70b-hf-4-3042-False-128-128-5374.6] PASSED                       [ 55%]
tests/test_text_generation_example.py::test_text_generation_fp8[token0-meta-llama/Llama-2-70b-hf-4-750-False-128-2048-7422.4] PASSED                       [ 61%]
tests/test_text_generation_example.py::test_text_generation_fp8[token0-meta-llama/Llama-2-70b-hf-4-207-False-2048-128-568.5] PASSED                        [ 66%]
tests/test_text_generation_example.py::test_text_generation_fp8[token0-meta-llama/Llama-2-70b-hf-8-172-False-2048-2048-4656.2] PASSED                      [ 72%]
tests/test_text_generation_example.py::test_text_generation_deepspeed[token0-meta-llama/Llama-2-70b-hf-8-1-64.10514998902435] PASSED                       [ 77%]
tests/test_text_generation_example.py::test_text_generation_deepspeed[token0-meta-llama/Meta-Llama-3-70B-Instruct-8-1-64] PASSED                           [ 83%]
tests/test_text_generation_example.py::test_text_generation_torch_compile[token0-meta-llama/Llama-2-7b-hf-102.27823420713148] PASSED                       [ 88%]
tests/test_text_generation_example.py::test_text_generation_torch_compile_distributed[token0-meta-llama/Llama-2-7b-hf-39.72973199515235] PASSED            [ 94%]
tests/test_text_generation_example.py::test_text_generation_distributed_tp[token0-meta-llama/Llama-2-7b-hf-1345.2369318328463] PASSED                      [100%]

========================================================= 18 passed, 34 deselected in 8202.18s (2:16:42) =========================================================

**Review by vidyasiv — APPROVED** [review]

**Comment by vidyasiv** [comment]
@regisss please take a look

**Review by regisss — COMMENTED** [review]
LGTM! There is a merge conflict to solve however before we can merge.

**Comment by zongwave** [comment]
> > LGTM! There is a merge conflict to solve however before we can merge.
> 
> @regisss @vidyasiv I merged the code changes with #948 and verified the CI test for #948 using command: GAUDI2_CI=1 RUN_SLOW=1 python -m pytest tests/transformers/tests/models/ -k test_generate_from_inputs_embeds_decoder_only
> 
> Unfortunately, 5 test cases failed. It seems this #948 requires generating new tokens from "input_ids" and "inputs_embeds" simultaneously. FAILED tests/transformers/tests/models/llama/test_modeling_llama.py::LlamaModelTest::test_generate_from_inputs_embeds_decoder_only - AssertionError: Lists differ: [[70,[66 chars]7, 78], [90, 71, 10, 82, 86, 98, 64, 64, 64, 6[37 chars] 64]] != [[70,[66 chars]7, 78, 39, 95, 41], [90, 71, 10, 82, 86, 98, 6[61 chars] 90]]
> 
> My PR only convers "inputs_embeds" or "input_ids" but not both at the same time. I’m curious about the real use case for this.
> 
> I need more time to make both two PRs work together.

@regisss @vidyasiv 

I **rebased the PR to the latest**, in this update, transformer models support process both "input_ids" and "inputs_embeds" at the same time or separately. 

**Passed** #948 test cases to **verify the generation consistent by input tokens or embeds** 
root@800671-4:/workspace/src/optimum-habana# GAUDI2_CI=1 RUN_SLOW=1 python -m pytest tests/transformers/tests/models/ -k test_generate_from_inputs_embeds_decoder_only
==================================================================== test session starts =====================================================================
platform linux -- Python 3.10.12, pytest-8.3.3, pluggy-1.5.0
rootdir: /workspace/src/optimum-habana
configfile: setup.cfg
plugins: anyio-4.4.0
collected 1226 items / 1218 deselected / 8 selected

tests/transformers/tests/models/bert/test_modeling_bert.py .                                                                                           [ 12%]
tests/transformers/tests/models/falcon/test_modeling_falcon.py .                                                                                       [ 25%]
tests/transformers/tests/models/gpt2/test_modeling_gpt2.py .                                                                                           [ 37%]
tests/transformers/tests/models/gpt_neox/test_modeling_gpt_neox.py .                                                                                   [ 50%]
tests/transformers/tests/models/gptj/test_modeling_gptj.py .                                                                                           [ 62%]
tests/transformers/tests/models/llama/test_modeling_llama.py .                                                                                         [ 75%]
tests/transformers/tests/models/roberta/test_modeling_roberta.py .                                                                                     [ 87%]
tests/transformers/tests/models/t5/test_modeling_t5.py .   

**Passed** test-text-generation CI test cases for single HPU device to **verify the performance reach benchmark** (can not reserve 8-card Gaudi2 passed the total cases before)
root@800671-4:/workspace/src/optimum-habana# GAUDI2_CI=1 RUN_SLOW=1 python -m pytest tests/test_text_generation_example.py -v -k llama
==================================================================== test session starts =====================================================================
platform linux -- Python 3.10.12, pytest-8.3.3, pluggy-1.5.0 -- /usr/bin/python
cachedir: .pytest_cache
rootdir: /workspace/src/optimum-habana
configfile: setup.cfg
plugins: anyio-4.4.0
collected 52 items / 34 deselected / 18 selected

tests/test_text_generation_example.py::test_text_generation_bf16_1x[token0-meta-llama/Llama-2-7b-hf-1-True-141.25776956002076] PASSED                  [  5%]
tests/test_text_generation_example.py::test_text_generation_bf16_1x[token0-meta-llama/Meta-Llama-3-8B-1-True-129] PASSED                               [ 11%]
tests/test_text_generation_example.py::test_text_generation_bf16_1x[token0-meta-llama/Llama-2-7b-hf-512-True-12808] PASSED                             [ 16%]
tests/test_text_generation_example.py::test_text_generation_bf16_1x[token0-meta-llama/Llama-2-7b-hf-512-False-8711] PASSED                             [ 22%]

File: optimum/habana/transformers/generation/utils.py [hunk 1]
```diff
@@ -922,34 +926,28 @@ def generate(
...
                         model_kwargs["inputs_embeds_offset"] = torch.tensor(
                             inputs_embeds_offset, device=inputs_tensor.device
                         )
-                        model_kwargs["inputs_embeds"] = torch.nn.functional.pad(
-                            model_kwargs["inputs_embeds"],
-                            (0, 0, 0, generation_config.max_new_tokens),
-                            value=generation_config.pad_token_id,
+
+                        model_kwargs["inputs_embeds_offset"] = torch.tensor(
+                            inputs_embeds_offset, device=inputs_tensor.device
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:937]: This is exactly the same as lines 932-933 right above, is this intended?

**Review by regisss — APPROVED** [review]

**Pull request merged**